contracts/StreamManager.vy

```diff
@@ -1,259 +1,464 @@
+flag Ability:
+    MODFIY_TOKENS
+    MODFIY_VALIDATORS
+    MODFIY_ACCESS
+    CANCEL_STREAMS
+
+capabilities: public(HashMap[address, Ability])
+
 
 event StreamCreated:
+    id: indexed(uint256)
+    owner: indexed(address)
     token: indexed(IERC20)
-    creator: indexed(address)
-    stream_id: indexed(uint256)
     amount_per_second: uint256
     start_time: uint256
-    reason: Bytes[MAX_REASON_SIZE]
+    products: DynArray[bytes32, MAX_PRODUCTS]
+
+
+event StreamOwnershipUpdated:
+    id: indexed(uint256)
+    old: indexed(address)
+    new: indexed(address)
 
 
 event StreamFunded:
-    creator: indexed(address)
-    stream_id: indexed(uint256)
-    amount_added: uint256
+    id: indexed(uint256)
+    funder: indexed(address)
+    added: uint256
 
 
-event StreamCancelled:
-    creator: indexed(address)
-    stream_id: indexed(uint256)
-    amount_locked: uint256
-    reason: Bytes[MAX_REASON_SIZE]
+event StreamClaimed:
+    id: indexed(uint256)
+    claimer: indexed(address)
+    exhausted: indexed(bool)
+    claimed: uint256
 
 
-event Claimed:
-    creator: indexed(address)
-    stream_id: indexed(uint256)
-    stream_exhausted: indexed(bool)
-    claimed_amount: uint256
+event StreamCancelled:
+    id: indexed(uint256)
+    canceller: indexed(address)
+    reason: indexed(bytes32)
+    refunded: uint256
 
 
 @deploy
 def __init__(
-    owner: address,
+    controller: address,
     min_stream_life: uint256,  # timedelta in seconds
-    validators: DynArray[Validator, MAX_VALIDATORS],
     accepted_tokens: DynArray[IERC20, 20],
+    validators: DynArray[Validator, MAX_VALIDATORS],
 ):
-    self.owner = owner
+    self.controller = controller
     MIN_STREAM_LIFE = min_stream_life
-    self.validators = validators
 
     for token: IERC20 in accepted_tokens:
         self.token_is_accepted[token] = True
 
+    self.validators = validators
+
 
 @external
-def set_validators(validators: DynArray[Validator, MAX_VALIDATORS]):
-    assert msg.sender == self.owner
-    self.validators = validators
+def transfer_control(new_controller: address):
+    """
+    @dev Begin the transfer of the `controller` role to `new_controller`.
+    @notice This action is very dangerous! Can only be performed by the `controller`. After the
+        transition is initiated, the `new_controller` must wait `CONTROLLER_ACCEPTANCE_DELAY`
+        before they are able to accept via `accept_control()`.
+    @param new_controller The address of the proposed new `controller` for this contract.
+    """
+    # NOTE: can revoke transfer at any time calling this method with `self.controller`
+    assert msg.sender == self.controller  # dev: not controller
+
+    log NewControllerProposed(msg.sender, new_controller)
+
+    self.new_controller = new_controller
+    self.new_controller_proposed = block.timestamp
 
 
 @external
-def add_token(token: IERC20):
-    assert msg.sender == self.owner
-    self.token_is_accepted[token] = True
+def accept_control():
```

Comment:
The foundational security property is that `controller` is the provider of some service, so if you feel your service would be interrupted by the change then you've 7 days to react to it

But doubtful, this is more of a utilitarian shift for a critical role in case you want to move wallets around. It is so critical though, hence 7 day revocation before acceptance